core/file_api.php

```diff
@@ -922,22 +967,66 @@ function file_ensure_uploaded( array $p_file ) {
 	switch( $p_file['error'] ) {
 		case UPLOAD_ERR_INI_SIZE:
 		case UPLOAD_ERR_FORM_SIZE:
-			trigger_error( ERROR_FILE_TOO_BIG, ERROR );
-			break;
+			throw new ClientException(
+				sprintf( "File '%s' too big", $p_file['name'] ),
+				ERROR_FILE_TOO_BIG );
+
 		case UPLOAD_ERR_PARTIAL:
 		case UPLOAD_ERR_NO_FILE:
-			trigger_error( ERROR_FILE_NO_UPLOAD_FAILURE, ERROR );
-			break;
-		default:
-			break;
+			throw new ClientException(
+				sprintf( "File '%s' upload failure", $p_file['name'] ),
+				ERROR_FILE_NO_UPLOAD_FAILURE );
 	}
 
 	if( ( '' == $p_file['tmp_name'] ) || ( '' == $p_file['name'] ) ) {
-		trigger_error( ERROR_FILE_NO_UPLOAD_FAILURE, ERROR );
+		throw new ClientException(
+			'File name or path is empty',
+			ERROR_FILE_NO_UPLOAD_FAILURE );
 	}
+
 	if( !is_readable( $p_file['tmp_name'] ) ) {
-		trigger_error( ERROR_UPLOAD_FAILURE, ERROR );
+		throw new ClientException( 'File is not readable', ERROR_UPLOAD_FAILURE );
+	}
+}
+
+/**
+ * Get mime type for the specified file.
+ *
+ * @param string $p_file_path The file path.
+ * @return boolean|string The mime type or false on failure.
+ */
+function file_get_mime_type( $p_file_path ) {
+	if( !file_exists( $p_file_path ) ) {
+		return false;
+	}
+
+	$t_info_file = config_get_global( 'fileinfo_magic_db_file' );
+
+	if( is_blank( $t_info_file ) ) {
+		$t_finfo = new finfo( FILEINFO_MIME );
+	} else {
+		$t_finfo = new finfo( FILEINFO_MIME, $t_info_file );
+	}
+
+	return $t_finfo->file( $p_file_path );
+}
+
+/**
+ * Get mime type for the specified content.
+ *
+ * @param string $p_file_path The file path.
+ * @return boolean|string The mime type or false on failure.
+ */
+function file_get_mime_type_for_content( $p_content ) {
+	$t_info_file = config_get_global( 'fileinfo_magic_db_file' );
+
+	if( is_blank( $t_info_file ) ) {
+		$t_finfo = new finfo( FILEINFO_MIME );
+	} else {
+		$t_finfo = new finfo( FILEINFO_MIME, $t_info_file );
 	}
```

Comment:
The code block starting with `$t_info_file =  ...` is the same like the block in function `file_get_mime_type`

Instead of the whole block, use existing function `finfo_get_if_available`.

Reply:
I just pushed a refactored version of the code.